Docs for proposed upgrade 4.1 (#331)
etherlinkcom/docs: docs/governance/how-do-i-participate-in-governance.mdx

1. The voting key claims the delegated voting rights by calling the `claim_voting_rights` entrypoint.

The delegation is not active until both of these steps are complete.

To propose delegation to a voting key, call the `propose_voting_key` entrypoint of the voting rights delegation contract.
By default, when you delegate your baking key to another key, that key receives the rights to vote for all governance contracts.
For example, the following command proposes delegation of voting rights for all governance contract from the key with the alias `my_wallet`.
It uses the variable `<MY_DELEGATE>` in place of the address of the account to delegate to:

```bash
octez-client transfer 0 from my_wallet to KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH \
  --arg "(Right (Pair \"<MY_DELEGATE>\" True None))" --burn-cap 0.1
```

Then, to claim the delegated voting rights, call the `claim_voting_rights` entrypoint from the delegated key and pass the baking key address.
For example, the following command claims the voting rights proposed in the previous command.
It uses the variable `<MY_BAKER>` in place of the address of the baking key address:

```bash

> octez-client transfer 0 from <MY_DELEGATE> to KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH \
>   --arg "(Left \"<MY_BAKER>\")" --burn-cap 0.1

Reviewer: ```suggestion
octez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH <MY_DELEGATE>\
  --entrypoint claim_voting_rights\
  --arg "\"<MY_BAKER>\""
```

```

> To remove the delegation and use your baking key to vote as before, propose delegation to the baking key and then claim the rights, as in this example:
> 

Reviewer: ```suggestion
To remove the delegation from your baker key, you can just recall propose_voting_key with (Pair False None), as in this example:

```

> 

Author: @charles-nl This doesn't seem to work. `propose_voting_key` accepts `pair address bool (option (set address))` so `(Pair False None)` doesn't fly. What's the best way for a baking key to revoke delegation?

```bash

> octez-client transfer 0 from <MY_BAKER> to KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH \
>   --arg "(Right (Pair \"<MY_BAKER>\" True None))" --burn-cap 0.1
> 
> octez-client transfer 0 from <MY_BAKER> to KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH \
>   --arg "(Left \"<MY_BAKER>\")" --burn-cap 0.1

Reviewer: ```suggestion
octez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH from my_wallet\
  --entrypoint propose_voting_key\
  --arg "(Pair \"<MY_DELEGATE>\" False None)" --burn-cap 0.1
```

```

You can also set the governance contracts to delegate rights for by specifying certain governance contracts.
The following example passes `True` and the addresses of the slow kernel governance and sequencer governance contracts as a proposal.
As a result, the delegated key can vote on those contracts but not on the kernel fast governance contract:

```bash

> octez-client transfer 0 from <MY_BAKER> to KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH \

Reviewer: ```suggestion
octez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH from <MY_BAKER>\
  --entrypoint propose_voting_key\
```

  --arg "(Pair \"<MY_DELEGATE>\" True (Some { \"KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf\" ; \"KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r\" }))"
```

Then, run this command to claim the voting rights for the specified contracts, which is just like the previous command to claim voting rights:

```bash

> octez-client transfer 0 from <MY_DELEGATE> to KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH \
>   --arg "(Left \"<MY_BAKER>\")" --burn-cap 0.1

Reviewer: ```suggestion
ctez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH <MY_DELEGATE>\
  --entrypoint claim_voting_rights\
  --arg "\"<MY_BAKER>\""
```

```

The following example passes `False` and the address of the fast kernel governance contract.
The result is the same as the previous proposal example: the delegated key can vote on the slow kernel governance and sequencer governance contracts but not the fast kernel governance contract.
You can also use this proposal to remove delegation for specific contracts:

```bash

> octez-client transfer 0 from <MY_BAKER> to KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH \
>   --arg "(Right (Pair \"<MY_DELEGATE>\" False (Some { \"KT1QDgF5pBkXEizj5RnmagEyxLxMTwVRpmYk\" })))" --burn-cap 0.1

Reviewer: ```suggestion
octez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH from <MY_BAKER>\
  --entrypoint propose_voting_key\
  --arg "(Pair \"<MY_DELEGATE>\" False (Some { \"KT1QDgF5pBkXEizj5RnmagEyxLxMTwVRpmYk\" }))" --burn-cap 0.1
```

<table class="customTableContainer">
<thead>
<tr>
<th>Contract</th>
<th>Mainnet address</th>
</tr>
</thead>
<tbody>
<tr>
<td>Kernel governance (slow)</td>
<td><InlineCopy code="KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r" href="https://better-call.dev/mainnet/KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r" abbreviate="6,4"></InlineCopy></td>
</tr>
<tr>
<td>Kernel governance (fast)</td>
<td><InlineCopy code="KT1DxndcFitAbxLdJCN3C1pPivqbC3RJxD1R" href="https://better-call.dev/mainnet/KT1DxndcFitAbxLdJCN3C1pPivqbC3RJxD1R" abbreviate="6,4"></InlineCopy></td>
</tr>
<tr>
<td>Sequencer operator</td>
<td><InlineCopy code="KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf" href="https://better-call.dev/mainnet/KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf" abbreviate="6,4"></InlineCopy></td>
</tr>
<tr>
<td>Voting rights delegation</td>
<td><InlineCopy code="KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH" href="https://better-call.dev/mainnet/KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH" abbreviate="6,4"></InlineCopy></td>
</tr>
<tr>
<td>Etherlink Smart Rollup</td>
<td><InlineCopy code="sr1Ghq66tYK9y3r8CC1Tf8i8m5nxh8nTvZEf" href="https://tzkt.io/sr1Ghq66tYK9y3r8CC1Tf8i8m5nxh8nTvZEf" abbreviate="6,4"></InlineCopy></td>
</tr>
</tbody>
</table>

> ## Delegating your key

Reviewer 2: ```suggestion
## Setting up a voting key
```

Or something like that.

Per a discussion between @germanD and @yurug we shouldn’t use the concept of delegate here as it would be confusing with Tezos L1 terminology.

Author: I disagree on this point because this is what "delegation" means: giving your right to vote to someone else. I think creating the concept of a voting key would cause more confusion than re-using delegation here.

```bash
octez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH from <MY_BAKER> \
  --entrypoint propose_voting_key \
  --arg "(Pair \"<MY_DELEGATE>\" True (Some { \"KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf\" ; \"KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r\" }))"
```

Then, run this command to claim the voting rights for the specified contracts, which is just like the previous command to claim voting rights:

```bash
octez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH <MY_DELEGATE> \
  --entrypoint claim_voting_rights \
  --arg "\"<MY_BAKER>\""
```

The following example passes `False` and the address of the fast kernel governance contract.
The result is the same as the previous proposal example: the voting key can vote on the slow kernel governance and sequencer governance contracts but not the fast kernel governance contract.
You can also use this proposal to remove delegation for specific contracts:

```bash
octez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH from <MY_BAKER> \
  --entrypoint propose_voting_key \
  --arg "(Pair \"<MY_DELEGATE>\" False (Some { \"KT1QDgF5pBkXEizj5RnmagEyxLxMTwVRpmYk\" }))" --burn-cap 0.1
```

:::tip

> The source code for the delegation contract is in the file [`delegated_governance.mligo`](https://gitlab.com/tezos/tezos/-/blob/master/etherlink/tezos_contracts/governance/contracts/delegated_governance.mligo?ref_type=heads) in the Octez repository.

Reviewer 3: ```suggestion
The source code for the delegation contract is in the file [`delegated_governance.mligo`](https://gitlab.com/tezos/tezos/-/blob/master/etherlink/tezos_contracts/governance/contracts/delegated_governance.mligo) in the Octez repository.
```
Super non-blocking, feel free to discard.

You can install LIGO and use its `compile expression` command and information from the source code to compile parameters to call it.

For example, from the code of the contract you can see that the parameter to pass to the `propose_voting_key` entrypoint is the type `address * bool * (address set) option`.
This command compiles an expression of this CameLIGO type to Michelson to propose delegation rights for two contracts:

```bash

> ligo compile expression cameligo "(\"<MY_DELEGATE>\" : address), True, (Some (Set.literal [(\"KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf\" : address); (\"KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r\" : address)]) : (address set) option)"

Reviewer 3: A proposition to use delimiting single-quotes in order to avoid having to escape double-quotes.
```suggestion
ligo compile expression cameligo '("<MY_DELEGATE>" : address), True, (Some (Set.literal [("KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf" : address); ("KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r" : address)]) : (address set) option)'
```

```

You can escape the double quotes in the result and use it as the parameter to pass to the `propose_voting_key` entrypoint.
Here is the result of the command:

```michelson
(Pair "<MY_DELEGATE>"
      True
      (Some { "KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf" ;
              "KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r" }))
```

Here is the resulting `octez-client` command:

```bash
octez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH from <MY_BAKER> \
  --entrypoint propose_voting_key \

>   --arg "(Pair \"<MY_DELEGATE>\" True (Some { \"KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r\" ; \"KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf\" }))"

Reviewer 3: ```suggestion
  --arg '(Pair "<MY_DELEGATE>" True (Some { "KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r" ; "KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf" }))'
```

```
:::

### Verifying voting rights

To check if an address has voting rights for a governance contract, pass the baking account address, the voting account address, and an option containing the governance contracts (or `None` for all contracts) to the `is_voting_key_of` view.
For example, the following command verifies that the key represented by the variable `<MY_DELEGATE>` has the rights to vote for the key represented by the variable `<MY_BAKER>`.
It returns True if the `<MY_BAKER>` key has delegated rights to the `<MY_DELEGATE>` key for all contracts that use this contract for delegation or False if it has not.

```bash
octez-client run view is_voting_key_of on contract KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH \

>   with input "Pair \"<MY_DELEGATE>\" \"<MY_BAKER>\" None"

Reviewer 3: ```suggestion
  with input 'Pair "<MY_DELEGATE>" "<MY_BAKER>" None'
```

```

To check delegation rights for a specific contract, pass the address as an option.
For example, this command verifies delegation for the slow kernel governance contract:

```bash
octez-client run view is_voting_key_of on contract KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH \

>   with input "Pair \"<MY_DELEGATE>\" \"<MY_BAKER>\" (Some \"KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r\")"

Reviewer 3: ```suggestion
  with input 'Pair "<MY_DELEGATE>" "<MY_BAKER>" (Some "KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r")'
```

The view accepts a tuple that includes the address of the voting key, the address of the baker key, and an option that is either None or a list of contracts.

For example, to compile the parameter from the previous example, install LIGO and run this command:

```bash
ligo compile expression cameligo "(\"<MY_DELEGATE>\" : address), (\"<MY_BAKER>\" : address), (Some (\"KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r\" : address): address option)"
```

You can escape the double quotes in the result and use it as the parameter to pass to the view.
Here is the result of the command:

```michelson
(Pair "tz1fsVnw7VQD73kUDB8ZWc67GWvCjTEibi9A"
      "tz1QCVQinE8iVj1H2fckqx6oiM85CNJSK9Sx"
      (Some "KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r"))
```

:::

### Ending delegated rights

To remove the delegation from your voting key, call the `propose_voting_key` entrypoint with the parameter `(Pair <MY_DELEGATE> False None)`, as in this example:

```bash
octez-client call KT1Ut6kfrTV9tK967tDYgQPMvy9t578iN7iH from my_wallet \
  --entrypoint propose_voting_key \

>   --arg "(Pair \"<MY_DELEGATE>\" False None)" --burn-cap 0.1

Reviewer 3: ```suggestion
  --arg '(Pair "<MY_DELEGATE>" False None)' --burn-cap 0.1
```

After a proposal wins a vote, any user can trigger the kernel upgrade by calling the appropriate governance contract's `trigger_kernel_upgrade` entrypoint:

```bash
octez-client transfer 0 from my_wallet to KT1DxndcFitAbxLdJCN3C1pPivqbC3RJxD1R \
  --entrypoint "trigger_kernel_upgrade" \
  --arg '"sr1Ghq66tYK9y3r8CC1Tf8i8m5nxh8nTvZEf"'
```

The command takes these parameters:

> - The address or Octez client alias of your account

Reviewer: ```suggestion
- The address or Octez client alias of your baker account or voting key
```

> - The address or Octez client alias of your account

Author: I see, only keys with baking rights can trigger the upgrade. Corrected.

When a proposal is in the Promotion period, you can vote for or against it or pass on voting by calling the `vote` entrypoint of the appropriate governance contract:

```bash
octez-client call KT1VZVNCNnhUp7s15d9RsdycP7C1iwYhAQ8r from my_wallet \
  --entrypoint "vote" --arg '"yea"'
```

The command takes these parameters:

- The address or Octez client alias of your baker account or the account to which you have delegated your Etherlink voting rights
- The address of the Etherlink kernel governance contract
- `"yea"`, `"nay"`, or `"pass"`, including the double quotes

The command does not need the hash of the kernel because only one kernel can be in the Promotion period at a time, so the options are to vote for or against that kernel or to abstain by voting "pass."

### Triggering upgrades

> After a proposal wins a vote, any account with voting rights can trigger the kernel upgrade by calling the appropriate governance contract's `trigger_kernel_upgrade` entrypoint:

Reviewer 3: Actually, that's really any user, you don't need to be a baker nor voting delegation to trigger a kernel upgrade that passed all voting phases, so I'd leave this line unchanged.

Author: Someone told me that you had to have baking rights -- I'll investigate.

When a proposal is in the Promotion period, you can vote for or against it or pass on voting by calling the `vote` entrypoint of the governance contract:

```bash
octez-client call KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf from my_wallet \
  --entrypoint "vote" --arg '"yea"'
```

The command takes these parameters:

- The address or Octez client alias of your baker account or the account to which you have delegated your Etherlink voting rights
- The address of the sequencer governance contract
- `"yea"`, `"nay"`, or `"pass"`, including the double quotes

For example:

```bash
octez-client call KT1WckZ2uiLfHCfQyNp1mtqeRcC1X6Jg2Qzf from tz1RLPEeMxbJYQBFbXYw8WHdXjeUjnG5ZXNq \
  --entrypoint "vote" --arg "\"yea\""
```

### Triggering sequencer operator updates

> After a proposed account wins a vote, any account with voting rights can trigger the change and enable that account to run the sequencer by calling the governance contract's `trigger_committee_upgrade` entrypoint:

Reviewer 3: As in another comment, it should be any user unless I missed something.
